src/otk/transform.py

```diff
@@ -112,15 +112,45 @@ def is_directive(needle: Any) -> bool:
     return isinstance(needle, str) and needle.startswith(PREFIX)
 
 
-@tree.must_be(dict)
-def define(ctx: Context, tree: Any) -> Any:
-    """Takes an `otk.define` block (which must be a dictionary and registers
-    everything in it as variables in the context."""
-
-    for key, value in tree.items():
-        ctx.define(key, value)
+def process_defines(ctx: Context, state: State, tree: Any):
+    """
+    Processes tree for new defines, resolving any references to other variables,
+    and update the global context. The State holds a reference to the nested
+    defines block the function is working in. New defines are added to the
+    nested block but references are resolved from the global ctx.defines.
+    """
 
-    return tree
+    subblock = state.defines
+
+    # Iterate over a copy of the tree so that we can modify it in-place.
+    for key, value in tree.copy().items():
+        if key.startswith("otk.define"):
+            # nested otk.define: process the nested values directly
+            process_defines(ctx, state, value)
+            continue
+
+        if key.startswith("otk.include"):
+            # TODO: disallow this (see https://github.com/osbuild/otk/issues/116)
```

Comment:
I guess the plan here is that we introduce it and then revert in a single commit so that (if we actually change our mind) we can get it back trivially ? Did I get this right?

Reply:
Yes, mostly.  But also my thinking was that this is a spec change that came from a discussion while the implementation was happening.  I wanted to isolate the change so that it has more visibility and stands on its own (e.g. a PR that only changes this, linked from the issue, that can be read on its own).

Also, I wasn't sure at the time if it needed some state tracking (something to tell us that we're in a defines block).  Now I realise since defines can't branch out to any other directives (except other defines), it's not really necessary.  But we should also think about what happens with other directives in general.